### ISSUE TITLE
Ambiguous EntityId as URL Parameter should probably be unambiguous and also work in headers

### ISSUE THREAD

**Issue by iJungleboy** [oqtane/oqtane.framework#1367]
From what I understand, we have security Attributes checking policies like these:

* User may edit Module
* User may edit Page
* User may view Module / Page

It is used like this:

```
    [Authorize(Policy = PolicyNames.ViewPage)]
```

Looking at the internals, it appears to do the following

1. Get an EntityId from the QueryString
2. Look up the page OR the Module based on that EntityId

I believe this has some weaknesses I would like to discuss:

1. It's not possible to check for both page and module permissions - for example requiring view-page and edit-module, since only one ID can be used which is always called `entityid`
2. There is a risk of the wrong ID being passed around - so a call could forward the PageId in the `entityid` but the endpoint may want to check for `PolicyNames.ViewModule` using the PageId - which of course would go undetected
3. In API scenarios it would often be helpful to allow using the parameters in headers instead of the URL

I recommend

1. Changing the parameter name from `entityid` to `moduleid` and `pageid` and ensure that these are used in the security checks to make them unambiguous
2. Change the Policy-check to use the values from the headers OR the url, with url as first preference as that's the one a user would see if ever he would see it

**Comment by iJungleboy** [comment]
@sbwalker thx for the detailed answer. So far we were able to work with all this without big problems - again because we figured out how it works and got it to fly. These are the main points I see to still recommend changing:

1. As of now, it seems that page permissions don't reflect in the module permissions. So a user having edit on the page will bounce on an API which requires edit on the module. I could be wrong, but that was my impression. 
2. If a developer creates a standalone solution which only affects his modules, then you are right - he knows what each endpoint will check and will supply the matching entity-id. But if the developer creates a library for getting something done -  resulting in APIs for use in other modules - then it becomes less clear. More parties have to do everything right for things to just work. If on the other hand API calls just supply `pageid` and `moduleid` and maybe someday other IDs as well, then the backend has the full context and can choose which ones it needs for checks. An updated implementation may extend to further security checks which previously didn't do anything - without breaking the "clients" which someone else made. 

Basically the "give the context, let the backend take what it needs" is closer to what Dnn did so far.

**Comment by sbwalker** [comment]
"Basically the "give the context, let the backend take what it needs" is closer to what Dnn did so far." - I actually purposely chose to go in different direction than DNN in this area. I always felt that permissions were a cross-cutting concern and should have been implemented as a centralized service in DNN - however DNN could not easily be adapted in this direction because of all of legacy baggage. In Oqtane I had the freedom to develop this area using a new approach.

In regards to the page/module permissions dependency, this is another deviation in behavior between DNN and Oqtane. In DNN there is a hierarchical relationship between Pages and Modules where some permissions are inherited from parent to child. This dependency caused a lot of challenges over the life of DNN so I did not opt for this approach in Oqtane. Page and Module permissions are completely independent from one another in Oqtane.

**Comment by iJungleboy** [comment]
I'm not sure I understand this approach

* Does this mean someone with page-edit permissions will never have permissions to edit the modules on it? 
* Or does this mean to give someone permissions on a page you would have to manually set them on each module as well?
* Or should the API check what pages the module is on (since it's not included in the request and a module can probably be on multiple pages), and also check if edit permissions were granted there?
* Or is this issue something for the future roadmap?

**Comment by sbwalker** [comment]
The goal in Oqtane is that the permissions for an Entity should be explicitly stored with that entity and should not be dependent on any other entity. So if you want to check if someone has the rights to Edit a Module ( whether in the Client or Server ) - you should simply check the Module Edit permissions. The implication is that if you want to have a logic dependency between two entities in your custom application then you need to ensure you explicitly set the permissions of the child entity when the permissions of the parent entity are modified. As I said, this is purposely different than DNN which attempted to implement a physical dependency between Pages and Modules for some permissions but not others - which caused a lot of confusion and maintenance headaches.

**Comment by sbwalker** [comment]
In regards to your other points about the passing of a singular entityid between the client and server, I think you have a valid point. I think this could be solved in a backward compatible manner with some fairly simple changes:

A new overload for CreateAuthorizationPolicyUrl() which allows a developer to pass multiple entities:

CreateAuthorizationPolicyUrl(string url, Dictionary<string, int> args)

used like:

CreateAuthorizationPolicyUrl(url, new Dictionary<string, int>() { { EntityNames.Page, 1 }, { EntityNames.Module, 2 } });

Producing a url with querystring parameters which are not likely to collide with other module behavior. I think header support would be a good solution here as well, as you suggested.

url?pageauthid=1&moduleauthid=2

A new property in ModuleControllerBase which extracts the querystring parameters into a Dictionary<string, int> for use in validation within the module logic.

protected Dictionary<string, int> _entity;

used like

if (_entity[EntityNames.Module] = id)

And a simple modification to the HandleRequirementAsync handler to recognize the new querystring parameters ( and still handle the legacy entityid ):

        protected override Task HandleRequirementAsync(AuthorizationHandlerContext context, PermissionRequirement requirement)
        {
            // permission is scoped based on EntityId which must be passed as a querystring parameter
            var ctx = _httpContextAccessor.HttpContext;
            if (ctx != null)
            {
                int entityId = -1;
                if (ctx.Request.Query.ContainsKey(requirement.EntityName + "authid"))
                {
                    entityId = int.Parse(ctx.Request.Query[requirement.EntityName + "authid"]);
                }
                // for backward compatibility
                if (ctx.Request.Query.ContainsKey("entityid"))
                {
                    entityId = int.Parse(ctx.Request.Query["entityid"]);
                }
                if (entityId != -1 && _userPermissions.IsAuthorized(context.User, requirement.EntityName, entityId, requirement.PermissionName))
                {
                    context.Succeed(requirement);
                }
                else
                {
                    _logger.Log(LogLevel.Error, this, LogFunction.Security, "User {User} Does Not Have {PermissionName} Permission For {EntityName}:{EntityId}", context.User, requirement.PermissionName, requirement.EntityName, entityId);
                }
            }
            return Task.CompletedTask;
        }

**Comment by iJungleboy** [comment]
Sounds 😎

Not sure yet if the addition of `authid` helps - it's certainly makes in unambiguous, but could add new confusion whereas `pageid` or `moduleid` would be simpler. But I understand that you would like to keep it separate in case certain calls may have a moduleid for other purposes, like a call to change a module-title (which may originate from Module A but target Module B... - I would prefer the `pageId` and `moduleId` but I guess it's architects-choice 👍

Note also to use `TryParse`

**Comment by sbwalker** [comment]
@iJungleboy I agree that "moduleid" is less ambiguous; however you need to consider that most modules developed in Oqtane will manage their own data and will be scoped to the module level. What this means is that they will most likely contain their own GET method which will need to retrieve a collection of data based on a moduleid. For example, this is a controller method generated by the Module Creator template:

```
        // GET: api/<controller>?moduleid=x
        [HttpGet]
        [Authorize(Policy = PolicyNames.ViewModule)]
        public IEnumerable<Models.DemoModule> Get(string moduleid)
        {
            return _DemoModuleRepository.GetDemoModules(int.Parse(moduleid));
        }

        // GET api/<controller>/5
        [HttpGet("{id}")]
        [Authorize(Policy = PolicyNames.ViewModule)]
        public Models.DemoModule Get(int id)
        {
            Models.DemoModule DemoModule = _DemoModuleRepository.GetDemoModule(id);
            if (DemoModule != null && DemoModule.ModuleId != _entityId)
            {
                DemoModule = null;
            }
            return DemoModule;
        }
```

You will notice that the first method uses a moduleid querystring parameter to retrieve a collection of records. It uses a querystring parameter so that it can disambiguate itself from the second GET method which retrieves a single record. However in both cases there is a need for the framework to pass an additional parameter for authorization purposes ( which is where "entityid" is currently being used ) - so you can see why this authorization parameter needs to have a non-conflicting name.

**Comment by sbwalker** [comment]
@iJungleboy after looking at the Blazor HttpClient it would require substantially more work and complexity to support passing these parameters as headers as opposed to continuing to use querystring parameters - so I am inclined to stick with querystring params for now.
